### PR TITLE
extend recycler verify mark to verify missing write barrier

### PR DESCRIPTION
**Pull request by leirocks**
record every write barrier bit set, while verifying mark, if an object is marked, verify the referencing pointer DID set the write barrier bit.
todo: support multiple recycler, clear the verification bits while destructing WriteBarrierPtr

and fix a bunch of missing write barrier found by this tool

### CODE REVIEW

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test OSX daily_osx_osx_debug_static  please

File: lib/Common/DataStructures/List.h [hunk 1]
```diff
@@ -360,6 +360,7 @@ namespace JsUtil
         void SetItem(int index, const T& item)
         {
             EnsureArray(index + 1);
+            // TODO: (SWB)(leish) find a way to force user defined copy constructor
```
Comment (jianchun) [inline lib/Common/DataStructures/List.h]: >// TODO: (SWB)(leish) find a way to force user defined copy constructor [](start = 12, length = 71)

This looks like "operator =", why copy constructor is involved? Could you please elaborate (possibly also update the comment)?

File: lib/Common/Memory/HeapBlock.h [hunk 2]
```diff
@@ -433,6 +433,7 @@ class SmallHeapBlockT : public HeapBlock
 #endif
     SmallHeapBlockBitVector* markBits;
     SmallHeapBlockBitVector  freeBits;
+    BVStatic<TBlockAttributes::PageCount * 0x1000 / sizeof(void*)> wbVerifyBits;
```
Comment (jianchun) [inline lib/Common/Memory/HeapBlock.h]: >BVStatic<TBlockAttributes::PageCount * 0x1000 / sizeof(void*)> wbVerifyBits; [](start = 4, length = 76)

1. Can this be in "#if DBG"?
2. Why the magic number "0x1000"? Could you add a comment?

File: lib/Common/Memory/RecyclerPointers.h [hunk 3]
```diff
@@ -310,8 +310,22 @@ class WriteBarrierPtr
     WriteBarrierPtr() : ptr(nullptr) {}
     WriteBarrierPtr(T * ptr)
     {
+#ifdef _WIN32
         // WriteBarrier
+        WriteBarrierSet(ptr);
+#else
+        // TODO: (leish)(swb) find a way to get dll load address on windows, and initialize card table for 
+        // image write copy region
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerPointers.h]: nit: comment confusing ("get ... on windows")

File: lib/Common/Memory/RecyclerWriteBarrierManager.h [hunk 4]
```diff
@@ -207,6 +207,8 @@ class RecyclerWriteBarrierManager
     static uint const s_BytesPerCardBit = 1 << s_BitArrayCardTableShift;  // 128 = 1 << 7
     static uint const s_BytesPerCard = s_BytesPerCardBit * 32;      // 4K  = 1 << 12 = 128 << 5
 
+    static Recycler* recycler; // assuming there's only one recycler
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerWriteBarrierManager.h]: >static Recycler* recycler; // assuming there's only one recycler [](start = 4, length = 64)

Could you put this and all related code under "#if DBG"?

Comment (leirocks) [inline lib/Common/Memory/RecyclerWriteBarrierManager.h]: I'm implementing list in Recycler so removing this

---
In reply to: [92254893](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92254893) [](ancestors = 92254893)

File: lib/JITIDL/ChakraJIT.idl [hunk 5]
```diff
@@ -3,8 +3,13 @@
 // Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
 //-------------------------------------------------------------------------------------------------------
 
+cpp_quote("#ifndef __JITTypes_h__")
+cpp_quote("#define __JITTypes_h__")
```
Comment (jianchun) [inline lib/JITIDL/ChakraJIT.idl]: A bit weird for this header to check that symbol and depend on definition in the other header. Likely able to do the same thing by only touching the other header ("JITTypes.h").

File: lib/JITIDL/JITTypes.h [hunk 6]
```diff
@@ -10,9 +10,19 @@ import "wtypes.idl";
...
+#define CHAKRA_WB_PTR void*
+#endif
 #define CHAKRA_PTR int
 #define BV_SHIFT 5
 #elif defined(_M_X64) || defined(_M_ARM64)
+#ifdef __midl
+#define CHAKRA_WB_PTR __int64
+#else
+#define CHAKRA_WB_PTR void*
+#endif
```
Comment (jianchun) [inline lib/JITIDL/JITTypes.h]: What's the reason for this change... Some fields need WB in the IDL headers structures? Because those structures were WB recycler allocated?

File: lib/Common/DataStructures/List.h (hunk 1 above)
Comment (leirocks) [inline lib/Common/DataStructures/List.h]: I implemented StackFrame& operator= for this, I meant if we have a way to force user defined copy constructor call, like 
this->buffer[index] = T::T(item);

it will be good, so we don't implement operator = for T. because the default operator = does not call write barrier on each field


---
In reply to: [92253745](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92253745) [](ancestors = 92253745)

File: lib/Jsrt/Jsrt.cpp [hunk 7]
```diff
@@ -3235,7 +3235,7 @@ JsErrorCode RunSerializedScriptCore(
 
         SourceContextInfo *sourceContextInfo;
         SRCINFO *hsi;
-        Js::FunctionBody *functionBody = nullptr;
+        Field(Js::FunctionBody*) functionBody = nullptr;
```
Comment (jianchun) [inline lib/Jsrt/Jsrt.cpp]: >Field(Js::FunctionBody*) functionBody = nullptr; [](start = 8, length = 48)

Why need this change? It's on stack...

File: lib/Runtime/Base/ScriptContext.h [hunk 8]
```diff
@@ -57,6 +57,44 @@ class SRCINFO
         SRCINFO* copySrcInfo = RecyclerNew(recycler, SRCINFO, *srcInfo);
         return copySrcInfo;
     }
+
+    SRCINFO()
+    {
+        memset(this, 0, sizeof(SRCINFO));
```
Comment (jianchun) [inline lib/Runtime/Base/ScriptContext.h]: Previously no constructor, why now needs "memset"?

File: lib/Runtime/Base/ScriptContext.h [hunk 9]
```diff
@@ -57,6 +57,44 @@ class SRCINFO
...
+        dlnHost(dlnHost),
+        ulColumnHost(ulColumnHost),
+        lnMinHost(lnMinHost),
+        ichMinHost(ichMinHost),
+        ichLimHost(ichLimHost),
+        ulCharOffset(ulCharOffset),
+        moduleID(moduleID),
+        grfsi(grfsi)
+    {
+    }
```
Comment (jianchun) [inline lib/Runtime/Base/ScriptContext.h]: Is this used anywhere?

File: lib/JITIDL/ChakraJIT.idl (hunk 5 above)
Comment (leirocks) [inline lib/JITIDL/ChakraJIT.idl]: previously Linux use JITTypes.h, windows uses the midl compiled result. without this change, every annotation in this file won't work on windows
with this change, both Linux and windows will use JITTypes.h. while RPC marshaling, it's using the compile result(chakraJIT.h), which does not contain any annotation.

---
In reply to: [92255395](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92255395) [](ancestors = 92255395)

File: lib/Common/DataStructures/List.h (hunk 1 above)
Comment (jianchun) [inline lib/Common/DataStructures/List.h]: Got you. Might the answer be impossible (language deficit? operator= and copy constructor are different no automatic behavior)? If so this comment is not useful.

---
In reply to: [92256073](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92256073) [](ancestors = 92256073,92253745)

File: lib/Runtime/Base/ScriptContext.h (hunk 9 above)
Comment (leirocks) [inline lib/Runtime/Base/ScriptContext.h]: there's such code:

            SRCINFO si = {
                /* sourceContextInfo   */ sourceContextInfo,
                /* dlnHost             */ 0,
                /* ulColumnHost        */ 0,
                /* lnMinHost           */ 0,
                /* ichMinHost          */ 0,
                /* ichLimHost          */ static_cast<ULONG>(scriptByteLength / chsize), // OK to truncate since this is used to limit sourceText in debugDocument/compilation errors.
                /* ulCharOffset        */ 0,
                /* mod                 */ kmodGlobal,
                /* grfsi               */ 0
            };

I changed one to use constructor. (the one found by tool). there more, might need to convert as well after the tool hit it



---
In reply to: [92257101](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92257101) [](ancestors = 92257101)

File: lib/Runtime/Base/ScriptContext.h (hunk 8 above)
Comment (leirocks) [inline lib/Runtime/Base/ScriptContext.h]: looks we don't it's allocated with RecyclerNewStructZ, let me remove this

---
In reply to: [92257043](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92257043) [](ancestors = 92257043)

File: lib/JITIDL/ChakraJIT.idl (hunk 5 above)
Comment (leirocks) [inline lib/JITIDL/ChakraJIT.idl]: midl does not happy with that

---
In reply to: [92259422](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92259422) [](ancestors = 92259422,92259202,92255395)

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
@dotnet-bot test Ubuntu daily_ubuntu_linux_release  please

**Comment by leirocks** [comment]
@dotnet-bot test Linux tests please

**Comment by leirocks** [comment]
implemented supporting multiple recyclers, clearing the bits, and for large heap blocks.
@jianchun @digitalinfinity @curtisman , can you take a look again?

File: lib/Common/Memory/Recycler.h [hunk 10]
```diff
@@ -2092,6 +2101,10 @@ class CollectedRecyclerWeakRefHeapBlock : public HeapBlock
 {
 public:
 #if DBG
+    virtual void WBSetBit(char* addr) override { Assert(false); }
+    virtual void WBSetBits(char* addr, uint length) override { Assert(false); }
+    virtual void WBClearBits(char* addr) override { Assert(false); }
```
Comment (jianchun) [inline lib/Common/Memory/Recycler.h]: WBClearBits signature looks inconsistent with the other two.

File: lib/Common/Memory/Recycler.h [hunk 11]
```diff
@@ -1921,6 +1921,15 @@ class Recycler
     } objectBeforeCollectCallbackState;
 
     bool ProcessObjectBeforeCollectCallbacks(bool atShutdown = false);
+
+#if DBG
+private:
+    static Recycler* recyclerList;
+    Recycler* next;
```
Comment (jianchun) [inline lib/Common/Memory/Recycler.h]: Just curious, do we use multiple recycler?

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp [hunk 12]
```diff
@@ -325,6 +335,11 @@ RecyclerWriteBarrierManager::WriteBarrier(void * address, size_t bytes)
     Assert(startIndex <= endIndex);
     memset(cardTable + startIndex, WRITE_BARRIER_PAGE_BIT | DIRTYBIT, endIndex - startIndex);
     GlobalSwbVerboseTrace(_u("Writing to 0x%p (CIndex: %u-%u)\n"), address, startIndex, endIndex);
+
+#if DBG
+    Recycler::WBSetBits((char*)address, (uint)bytes/sizeof(void*));
```
Comment (jianchun) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: >(uint) [](start = 40, length = 6)

Is this "(uint)" cast needed? Does it divide by "sizeof" type and again result in "size_t"?

File: lib/Backend/JITTimePolymorphicInlineCacheInfo.cpp [hunk 13]
```diff
@@ -53,8 +53,9 @@ JITTimePolymorphicInlineCacheInfo::InitializePolymorphicInlineCacheInfo(
     __in Js::PolymorphicInlineCacheInfo * runtimeInfo,
     __out PolymorphicInlineCacheInfoIDL * jitInfo)
 {
+#pragma warning(suppress: 6001)
```
Comment (leirocks) [inline lib/Backend/JITTimePolymorphicInlineCacheInfo.cpp]: inside operator=, there's *this, preFast though it's dereferencing and using an uninitialized memory, here itself is actually assignment(i.e. initializing)
before the change polymorphicCacheUtilizationArray is not wrapped with WriteBarrierPtr and does not have operator= and *this

---
In reply to: [92454774](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92454774) [](ancestors = 92454774)

File: lib/Common/Memory/Recycler.h (hunk 11 above)
Comment (leirocks) [inline lib/Common/Memory/Recycler.h]: just in case, there's multiple thread context we'll have multiple recycler, aka, node case?

---
In reply to: [92457813](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92457813) [](ancestors = 92457813)

File: lib/Common/Memory/RecyclerWriteBarrierManager.cpp (hunk 12 above)
Comment (leirocks) [inline lib/Common/Memory/RecyclerWriteBarrierManager.cpp]: bytes is size_t

---
In reply to: [92458202](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92458202) [](ancestors = 92458202)

File: lib/Backend/JITTimePolymorphicInlineCacheInfo.cpp (hunk 13 above)
Comment (leirocks) [inline lib/Backend/JITTimePolymorphicInlineCacheInfo.cpp]: sure, I can do it in next commit

---
In reply to: [92459896](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92459896) [](ancestors = 92459896,92458452,92454774)

Comment (jianchun) [inline lib/Backend/JITTimePolymorphicInlineCacheInfo.cpp]: Seems strange. "WriteBarrierPtr" has default constructor which initialize ptr to null, thus inside "operator=" there is no uninitialized memory.

File: lib/Runtime/Base/FunctionBody.h [hunk 14]
```diff
@@ -3690,7 +3690,7 @@ namespace Js
         static uint const ScopeMetadataSlotIndex = 1;    // Either a FunctionBody* or DebuggerScope*
         static uint const FirstSlotIndex = 2;
     public:
-        ScopeSlots(Var* slotArray) : slotArray(slotArray)
+        ScopeSlots(Var* slotArray) : slotArray((Field(Var)*)slotArray)
```
Comment (ThomsonTan) [inline lib/Runtime/Base/FunctionBody.h]: >(Field(Var)* [](start = 47, length = 12)

Any difference between `Field(Var)*` and `Field(Var*)`?

Comment (leirocks) [inline lib/Runtime/Base/FunctionBody.h]: Field(Var)* is WriteBarrierPtr<T*> *,  Field(Var) is WriteBarrierPtr<T**>
weird the comment does not show the star after '>'

---
In reply to: [92464727](https://github.com/Microsoft/ChakraCore/pull/2191#discussion_r92464727) [](ancestors = 92464727,92464304)
